File: api/server/book/internal/application/input/book.go
package input

// ListOnBookshelf - 本棚ないの書籍一覧取得のリクエスト
type ListBookshelf struct {
	UserID string `json:"userId" validate:"required"`
	Limit  int    `json:"limit" validate:"gte=0,lte=1000"`
	Offset int    `json:"offset" validate:"gte=0"`
}

// Book - 書籍登録/更新のリクエスト
type Book struct {
	Title          string        `json:"title" validate:"required,max=64"`
	TitleKana      string        `json:"titleKana" validate:"required,max=128"`
	Description    string        `json:"description" validate:"omitempty,max=2000"`
	Isbn           string        `json:"isbn" validate:"required,max=16"`
	Publisher      string        `json:"publisher" validate:"required,max=32"`
	PublishedOn    string        `json:"publishedOn" validate:"required"`
	ThumbnailURL   string        `json:"thumbnailUrl" validate:"omitempty"`
	RakutenURL     string        `json:"rakutenUrl" validate:"omitempty"`
	RakutenGenreID string        `json:"rakutenGenreId" validate:"omitempty"`
	Authors        []*BookAuthor `json:"authors" validate:"omitempty"`
}

// BookAuthor - 著者のリクエスト
type BookAuthor struct {
	Name     string `json:"name" validate:"required,max=32"`
	NameKana string `json:"nameKana" validate:"required,max=64"`
}

// Bookshelf - 本棚への書籍登録/更新のリクエスト
type Bookshelf struct {
	UserID string `json:"userId" validate:"required"`
	BookID int    `json:"bookId" validate:"required,gte=1"`
	Status int    `json:"status" validate:"required,gte=0,lte=5"`
	ReadOn string `json:"readOn" validate:"omitempty"`
}

File: api/gateway/native/src/types/output/book.ts
/*
 * 書籍詳細アウトプット
 */
export interface IBookOutput {
  id: number
  title: string
  titleKana: string
  description: string
  isbn: string
  publisher: string
  publishedOn: string
  thumbnailUrl: string
  rakutenUrl: string
  rakutenSize: string
  rakutenGenreId: string
  createdAt: string
  updatedAt: string
  authors: Array<IBookOutputAuthor>
}

export interface IBookOutputAuthor {
  name: string
  nameKana: string
}

/*
 * 書籍一覧(ハッシュ)アウトプット
 */
export interface IBookHashOutput {
  [key: number]: IBookHashOutputBook
}

export interface IBookHashOutputBook {
  id: number
  title: string
  titleKana: string
  description: string
  isbn: string
  publisher: string
  publishedOn: string
  thumbnailUrl: string
  rakutenUrl: string
  rakutenSize: string
  rakutenGenreId: string
  createdAt: string
  updatedAt: string
  authors: Array<IBookOutputAuthor>
}

export interface IBookHashOutputAuthor {
  name: string
  nameKana: string
}

/*
 * 本棚の書籍詳細アウトプット
 */
export interface IBookshelfOutput {
  id: number
  bookId: number
  userId: string
  status: number
  readOn: string
  createdAt: string
  updatedAt: string
  book: IBookshelfOutputBook
  myReview: IBookshelfOutputReview
}

export interface IBookshelfOutputBook {
  id: number
  title: string
  titleKana: string
  description: string
  isbn: string
  publisher: string
  publishedOn: string
  thumbnailUrl: string
  rakutenUrl: string
  rakutenSize: string
  rakutenGenreId: string
  createdAt: string
  updatedAt: string
  authors: Array<IBookshelfOutputAuthor>
  reviews: Array<IBookshelfOutputReview>
}

export interface IBookshelfOutputAuthor {
  name: string
  nameKana: string
}

export interface IBookshelfOutputReview {
  id: number
  userId: string
  score: number
  impression: string
  createdAt: string
  updatedAt: string
}

/*
 * 本棚の書籍一覧アウトプット
 */
export interface IBookshelfListOutput {
  bookshelves: Array<IBookshelfListOutputBookshelf>
  limit: number
  offset: number
  total: number
}

export interface IBookshelfListOutputAuthor {
  name: string
  nameKana: string
}

export interface IBookshelfListOutputBook {
  id: number
  title: string
  titleKana: string
  description: string
  isbn: string
  publisher: string
  publishedOn: string
  thumbnailUrl: string
  rakutenUrl: string
  rakutenSize: string
  rakutenGenreId: string
  createdAt: string
  updatedAt: string
  authors: Array<IBookshelfListOutputAuthor>
}

export interface IBookshelfListOutputBookshelf {
  id: number
  bookId: number
  userId: string
  status: number
  readOn: string
  createdAt: string
  updatedAt: string
  book: IBookshelfListOutputBook
  myReview: IBookshelfListOutputReview
}

export interface IBookshelfListOutputReview {
  id: number
  userId: string
  score: number
  impression: string
  createdAt: string
  updatedAt: string
}

/*
 * 書籍レビュー詳細アウトプット
 */
export interface IReviewOutput {
  id: number
  bookId: number
  userId: string
  score: number
  impression: string
  createdAt: string
  updatedAt: string
}

/*
 * 書籍レビュー一覧アウトプット
 */
export interface IReviewListOutput {
  reviews: Array<IReviewListOutputReview>
  limit: number
  offset: number
  total: number
  order?: IReviewListOutputOrder
}

export interface IReviewListOutputReview {
  id: number
  bookId: number
  userId: string
  score: number
  impression: string
  createdAt: string
  updatedAt: string
}

export interface IReviewListOutputOrder {
  by: string
  direction: string
}
